thin_client/thin_client_main.py

```diff
@@ -0,0 +1,111 @@
+    pygame.key.set_repeat(frameInterval, frameInterval) # 1 input per frame
+    
+    myFont = pygame.font.Font(None, 30)
+    label = myFont.render("Loading...", True, (255, 255, 255))
+    mouseLabel = myFont.render("Press ESC to lock/unlock mouse.", True, (255, 255, 255))
+    textRect = label.get_rect()
+    renderPosX = screen.get_rect().centerx - textRect.centerx
+    renderPosY = screen.get_rect().centery - textRect.centery
+    screen.blit(label, (renderPosX, renderPosY))
+    screen.blit(mouseLabel, (renderPosX - 100, renderPosY + 50))
+    pygame.display.update()
+
+    return pygame
+
+def processMouseButtons(playerControllerID, sequence, sock, pygame, event):
+    leftMouseButton, middleMouseButton, rightMouseButton = pygame.mouse.get_pressed()
+    if (leftMouseButton == 1):
+        UECharCode = 1
+    elif (middleMouseButton == 1):
+        UECharCode = 4
+    elif (rightMouseButton == 1):
+        UECharCode = 2
+    else:
+        UECharCode = 0
+    UEKeyCode = UECharCode
+    if (event.type == MOUSEBUTTONDOWN):
+        sequence = networking.packAndSend(s.DEVICE_KEYBOARD, sequence, playerControllerID, UEKeyCode, UECharCode, 2, sock)
+    elif (event.type == MOUSEBUTTONUP):
+        sequence = networking.packAndSend(s.DEVICE_KEYBOARD, sequence, playerControllerID, UEKeyCode, UECharCode, 3, sock)
+    print(pygame.mouse.get_pressed(), "=>", UEKeyCode)
+    
+def processMouseMotion(playerControllerID, sequence, sock, pygame, event):
+    x, y = pygame.mouse.get_rel()
+    sequence = networking.packAndSend(s.DEVICE_MOUSE, sequence, playerControllerID, x, y, event.type, sock)
+    print("Mouse:", x,y)
+
+def processKeyboardPress(playerControllerID, sequence, sock, pygame, event):
+    print("ASCII Key is:", event.key)
+    UEKeyCode = s.ASCII_TO_UE_KEYCODE.get(event.key, 0)
+    UECharCode = s.ASCII_TO_UE_CHARCODE.get(event.key, UEKeyCode)
+    UEKeyCode = UECharCode or UEKeyCode # This code is redundant. It changes nothing.
+    print(UEKeyCode, UECharCode)
+    sequence = networking.packAndSend(s.DEVICE_KEYBOARD, sequence, playerControllerID, UEKeyCode, UECharCode, event.type, sock)
+    print(event.key, "=>", UEKeyCode)
+
+def toggleMouseGrab(pygame, isMouseGrabbed):
+    if (isMouseGrabbed == True):
+        isMouseGrabbed = False
+        pygame.event.set_grab(False)
+        pygame.mouse.set_visible(True)
+    else:
+        isMouseGrabbed = True
+        pygame.event.set_grab(True)
+        pygame.mouse.set_visible(False)
+        
+    return isMouseGrabbed
+
+def startClient(IP, port, playerControllerID):
+    sequence = 0
+    
+    sock = networking.initializeUDPSocket()
+    pygame = initializePygame(s.FPS) #FPS
+    pygame = vlc_addon.initializeStream(IP, port, pygame)
+    isRunning = True
+    isMouseGrabbed = True
+
+    while (isRunning):
+        event = pygame.event.wait() # program will sleep if there are no events in the queue
+
+        if (event.type == KEYDOWN or event.type == KEYUP):
```

Comment:
Optionally, it can be written in object oriented (with Polymorphic behavior).

    if is_mouse():
        input_processor = MouseProcessor(...)
    elif is_keyboard():
        input_processor = KeyProcessor(...)
    input_processor.process()